Introduce raw byte arrays for psk hint and identity.

boaks opened the pull request:
Support PSK clients with non-compliant encoding.
Fixes issue #902 .

Signed-off-by: Achim Kraus <achim.kraus@bosch-si.com>

boaks commented:
I would like to merge this PR at the end of the week.

The implementation keeps the bytes and string and replaces the string with the string from the store, if it's not encoded compliant (no UTF-8). It's intended to support broken clients as work around until they are fixed.

sbernard31 commented on scandium-core/src/main/java/org/eclipse/californium/scandium/config/DtlsConnectorConfig.java:
```diff
@@ -1129,23 +1150,54 @@ public Builder setRetransmissionTimeout(int timeout) {
+		 * @return this builder for command chaining
+		 * @throws IllegalArgumentException if the store is already set with
+		 *             {@link #setPskStore(PskStore)}
+		 */
```
Or maybe now setPskStore is deprecated ? I think having both API could be confusing without explanation which should be used for what.

boaks replied:
Using the `BytesPskStore` is my preferred way, Using `PskStore` is just for backwards compatibility. I can add some javadoc.

sbernard31 replied:
So we should at least deprecate `PskStore` ?
(Personally, If you remove it I'm ok too)

boaks replied:
OK, I would go for deprecate and remove it after M15.

boaks replied:
Or copy the BytesPskStore into PskStore and remove the new one. I think that makes most sense.

sbernard31 commented on scandium-core/src/main/java/org/eclipse/californium/scandium/dtls/pskstore/BytesPskStore.java:
```diff
@@ -0,0 +1,112 @@
+	 * <p>
+	 * The implementation is intended to normalize the identity by a matching
+	 * entry, if that entry is not UTF-8 compliant encoded.
+	 * </p>
```
I spend some times to understand the idea behind normalizing. This is mainly for clearer log, right ?
Maybe we could explain more about that and maybe adding an example of value for pskPublicinformation for not UTF-8 compliant encoded in javadoc could help to better understand the idea.

boaks replied:
No, it's mainly for the principal `PreSharedKeyIdentity`, that's the major downside of "non-compliant" clients. The principal requires a proper, unique name.

sbernard31 replied:
OK, so this confirms this deserve more explanation :sweat_smile:

boaks replied:
I try it.

boaks commented on scandium-core/src/main/java/org/eclipse/californium/scandium/config/DtlsConnectorConfig.java:
```diff
@@ -475,15 +480,30 @@ public PrivateKey getPrivateKey() {
+	 * Gets the registry of <em>shared secrets</em> based on UTF-8 strings used
+	 * for authenticating clients during a DTLS handshake.
+	 * 
+	 * This store is used indirect by a {@link MappedBytesPskStore}. Non UTF-8
```
OK

boaks commented on scandium-core/src/main/java/org/eclipse/californium/scandium/config/DtlsConnectorConfig.java:
```diff
@@ -1129,23 +1150,54 @@ public Builder setRetransmissionTimeout(int timeout) {
 		 * {@link #setSupportedCipherSuites(String...)}.
 		 * 
-		 * @param pskStore the key store
+		 * This store is used indirect by a automatically created
```
OK

boaks commented on scandium-core/src/main/java/org/eclipse/californium/scandium/dtls/PskPublicInformation.java:
```diff
@@ -0,0 +1,198 @@
+	/**
+	 * Create PSK public information from bytes (identity or hint).
+	 * 
+	 * Used by
```
Yes, I added it.

boaks commented on scandium-core/src/main/java/org/eclipse/californium/scandium/dtls/PskPublicInformation.java:
```diff
@@ -0,0 +1,198 @@
+	}
+
+	/**
+	 * Create PSK public information from bytes (identity or hint).
```
yes

boaks commented:
> The main "issue" is just now I don't know which store/interface I should use.

I would have preferred to keep the `PskStore` as main API. But I can't see, how this works. Therefore I changed to the `BytesPskStore`.

boaks commented:
I will do the stuff on Monday.

boaks commented:
I removed the `BytePskStore` and moved the bytes-stuff into the `PskStore` (API break).
A `StringPskStore` is added for easier backwards compatibility.

sbernard31 commented:
> I removed the BytePskStore and moved the bytes-stuff into the PskStore (API break).
A StringPskStore is added for easier backwards compatibility.

:+1:   LGTM !
